### PR TITLE
✨ [RUMF-1175] collect reports and csp violation

### PR DESCRIPTION
**Pull request by amortemousque**
## Motivation

Collect reports and csp violation.
Feature under feature flag

Worth reviewing commit by commit 

## Changes

- add a reportObserbvable 
- collect error reports in RUM
- collect all reports in Logs

## Testing

- [x] Local
- [ ] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/core/src/domain/report/reportObservable.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+interface InterventionReportBody {
+  id: string
+  message: string
+  lineNumber: number
+  columnNumber: number
+  sourceFile: string
+}
+
+export const ReportType = {
```
Comment (amortemousque) [inline packages/core/src/domain/report/reportObservable.ts:48]: Can't be an enum because, it will be exposed with a new init config option

File: packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import { clocksNow, ErrorHandling, ErrorSource, initReportObservable, ReportType } from '@datadog/browser-core'
+
+export function trackReportError(errorObservable: Observable<RawError>) {
+  const subscription = initReportObservable([ReportType.csp_violation, ReportType.intervention]).subscribe(
+    (reportError) =>
+      errorObservable.notify({
+        startClocks: clocksNow(),
+        message: reportError.message,
+        stack: reportError.stack,
+        source: ErrorSource.REPORT,
```
Comment (amortemousque) [inline packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts:11]: The rum events format will need to be updated

File: packages/core/src/domain/report/reportObservable.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+export type ReportType = typeof ReportType[keyof typeof ReportType]
+
+export interface Report {
+  type: ReportType
+  message: string
+  stack?: string
+}
+
+export type BrowserReportType = 'intervention' | 'deprecation'
```
Comment (bcaudan) [inline packages/core/src/domain/report/reportObservable.ts:63]: **💭 thought:** ‏It is a bit hard to follow what is Browser types and what is our types. 
It could be interesting to extract the browser types to a dedicated file with the actual browser names somewhere.

Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts]: I've put browser types into report/browser.type.ts

File: packages/core/src/domain/report/reportObservable.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import type {
+  Report,
+  BrowserWindow,
+  ReportType,
+  ReportingObserverCallback,
+  ReportingObserverOption,
+  ReportingObserver as ReportingObserverInterface,
+} from './browser.types'
+
+declare const ReportingObserver: {
```
Comment (amortemousque) [inline packages/core/src/domain/report/reportObservable.ts:13]: I tried to move the ReportingObserver type declaration in the `browser.types` without success. I run into several typescript issues. This is the best solution I could found.

File: packages/core/src/domain/report/reportObservable.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  new (callback: ReportingObserverCallback, option: ReportingObserverOption): ReportingObserverInterface
```
Comment (bcaudan) [inline packages/core/src/domain/report/reportObservable.ts:14]: **💬 suggestion:** could you:

- move the constructor type in the existing ReportingObserver interface
- ensure to use ReportingObservable variable from the BrowserWindow type
- remove the declare const

File: packages/core/src/domain/report/reportObservable.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+export const CustomReportType = {
```
Comment (bcaudan) [inline packages/core/src/domain/report/reportObservable.ts:17]: **❓ question:** ‏what do you want to express with "custom"?

File: packages/core/src/domain/report/reportObservable.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  intervention: 'intervention',
+  deprecation: 'deprecation',
+  csp_violation: 'csp_violation',
```
Comment (bcaudan) [inline packages/core/src/domain/report/reportObservable.ts:20]: **💭 thought:** ‏we don't have a rule for that but this case seems odd compared to what we currently do.
I would have expected either:

- `CSP_VIOLATION: 'csp_violation'`
- `cspViolation: 'csp_violation'`

File: packages/core/src/domain/report/reportObservable.spec.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    reportingObserverStub = stubReportingObserver()
+    cspEventListenerStub = stubCspEventListener()
+    notifyReport = jasmine.createSpy('notifyReport')
+  })
+
+  afterEach(() => {
+    reportingObserverStub.reset()
+    consoleSubscription.unsubscribe()
+  })
+  ;[CustomReportType.deprecation, CustomReportType.deprecation].forEach((type) => {
```
Comment (BenoitZugmeyer) [inline packages/core/src/domain/report/reportObservable.spec.ts:26]: **🔨 warning:** ‏looks like it defines the same test twice

Comment (bcaudan) [inline packages/core/src/domain/report/reportObservable.spec.ts:26]: **🥜 nitpick:** ‏
```suggestion
  })
  
  ;[CustomReportType.deprecation, CustomReportType.deprecation].forEach((type) => {
```

Comment (bcaudan) [inline packages/core/src/domain/report/reportObservable.spec.ts:26]: **🥜 nitpick:** ‏
```suggestion
  })
  
  ;[CustomReportType.deprecation, CustomReportType.deprecation].forEach((type) => {
```

File: packages/core/src/domain/report/reportObservable.spec.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+describe(`report observable`, () => {
+  let reportingObserverStub: { reset(): void; raiseReport(type: string): void }
+  let cspEventListenerStub: { dispatchEvent(): void }
+  let consoleSubscription: Subscription
+  let notifyReport: jasmine.Spy
+
+  beforeEach(() => {
+    if (!isChromium()) {
+      pending('no ReportingObserver support')
+    }
```
Comment (BenoitZugmeyer) [inline packages/core/src/domain/report/reportObservable.spec.ts:15]: **❓ question:** Since we are stubing browser APIs for those tests, do we need to worry about whether the API is supported or not? I think it could be run in all browsers, no?

File: packages/core/src/domain/report/reportObservable.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  sourceFile: string | null,
+  lineNumber: number | null,
+  columnNumber: number | null
+): string | undefined {
+  if (!sourceFile) {
+    return undefined
+  }
+
+  return `${type}: ${message}
+at <anonymous> @ ${[sourceFile, lineNumber, columnNumber].filter((e) => e).join(':')}`
```
Comment (BenoitZugmeyer) [inline packages/core/src/domain/report/reportObservable.ts:118]: **💬 suggestion:** ‏maybe use [`toStackTraceString()`](https://github.com/DataDog/browser-sdk/blob/5f884b7b23de943d4b70b3ca192b4dfd455db8ca/packages/core/src/tools/error.ts#L63) here

File: packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.spec.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    clock = mockClock()
+  })
+
+  afterEach(() => {
+    subscription.unsubscribe()
+    clock.cleanup()
+    reportingObserverStub.reset()
+  })
+
+  it('should error report', () => {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.spec.ts:29]: **❓ question:** ‏should it not be enable only behind the FF?

File: packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts (hunk 2 above)
Comment (bcaudan) [reply packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts:12]: Should not it be done before this PR?

File: packages/logs/src/boot/startLogs.spec.ts [hunk 12]
```diff
@@ -202,6 +208,24 @@ describe('logs', () => {
...
     })
+
+    it('should send reports', () => {
+      if (!isChromium()) {
+        pending('no ReportingObserver support')
+      }
+      const logger = new Logger(noop)
+      const logErrorSpy = spyOn(logger, 'log')
+      const reportingObserverStub = stubReportingObserver()
+      updateExperimentalFeatures(['forward-reports'])
```
Comment (bcaudan) [inline packages/logs/src/boot/startLogs.spec.ts:219]: **💬 suggestion:** ‏what about adding a test without the FF to ensure that it also behave correctly?

File: packages/core/test/specHelper.ts [hunk 13]
```diff
@@ -417,3 +424,80 @@ export function stubCookie() {
     },
   }
 }
+
+export function stubReportingObserver() {
```
Comment (BenoitZugmeyer) [inline packages/core/test/specHelper.ts:428]: **💭 thought:** Maybe we'll need to split this file one day, it's beginning to be a bit large

File: packages/logs/src/boot/startLogs.spec.ts [hunk 14]
```diff
@@ -202,6 +208,24 @@ describe('logs', () => {
...
+      originalStartLogs({ ...baseConfiguration, forwardReports: ['intervention'] }, logger)
```
Comment (bcaudan) [inline packages/logs/src/boot/startLogs.spec.ts:220]: **💬 suggestion:**  it could be interesting to add a test for the expected default behavior as well‏

File: packages/core/src/domain/report/reportObservable.ts (hunk 6 above)
Comment (BenoitZugmeyer) [inline packages/core/src/domain/report/reportObservable.ts:17]: **❓ question:** ‏I wonder, why is it called `CustomReportType` and not simply `ReportType`?

Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:17]: I just don't want to conflict with native types that lives in `browser.types.ts`.

Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:17]: Same comment as https://github.com/DataDog/browser-sdk/pull/1332#discussion_r816747785

File: packages/core/test/specHelper.ts (hunk 13 above)
Comment (amortemousque) [reply packages/core/test/specHelper.ts:428]: I agree

File: packages/core/src/domain/report/reportObservable.spec.ts (hunk 9 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.spec.ts]: Yes, why not :)

File: packages/core/src/domain/report/reportObservable.ts (hunk 7 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts]: I see, let go with the` cspViolation: 'csp_violation'`

File: packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.spec.ts (hunk 11 above)
Comment (amortemousque) [reply packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.spec.ts]: Of course

File: packages/logs/src/boot/startLogs.spec.ts (hunk 14 above)
Comment (amortemousque) [reply packages/logs/src/boot/startLogs.spec.ts]: By the expected default behaviour, you mean without the FF?

File: packages/logs/src/boot/startLogs.spec.ts (hunk 12 above)
Comment (amortemousque) [reply packages/logs/src/boot/startLogs.spec.ts:236]: Sure

File: packages/core/src/domain/report/reportObservable.spec.ts (hunk 8 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.spec.ts]: This is the prettier issue reported here: https://github.com/DataDog/browser-sdk/blob/ae205749e656fb477c870625b075a6c341d09250/packages/core/src/domain/console/consoleObservable.spec.ts#L7-L8

File: packages/core/src/domain/report/reportObservable.spec.ts [hunk 15]
```diff
@@ -0,0 +1,63 @@
+import { isChromium } from '../../tools/browserDetection'
+import type { Subscription } from '../../tools/observable'
+import { stubReportingObserver, stubCspEventListener } from '../../../test/specHelper'
+import { initReportObservable, CustomReportType } from './reportObservable'
+
+describe(`report observable`, () => {
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.spec.ts]: I prepare the PR to enable the TS rule :)

File: packages/core/src/domain/report/browser.types.ts [hunk 16]
```diff
@@ -0,0 +1,43 @@
+export interface BrowserWindow {
+  ReportingObserver?: ReportingObserver
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/core/src/domain/report/browser.types.ts:2]: Yes sure

File: packages/core/src/domain/report/reportObservable.ts (hunk 5 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:14]: If I declare the ReportingObserver in browser.types like so:
```typescript
export interface BrowserWindow {
  ReportingObserver: ReportingObserver
}

declare const ReportingObserver: {
  prototype: PerformanceEntry
  new (callback: ReportingObserverCallback, option: ReportingObserverOption): ReportingObserver
}

interface ReportingObserver {
  disconnect(): void
  observe(): void
  takeRecords(): Report[]
}
```

TS complains when I try use ReportingObservable variable from the BrowserWindow type:
```typescript
const test = new (window as BrowserWindow).ReportingObserver()
```
TS error message:
`This expression is not constructable. Type 'ReportingObserver' has no construct signatures.`

Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:14]: Same if I declare a class like so:
```typescript
declare class ReportingObserver { 
  constructor(callback: ReportingObserverCallback, option: ReportingObserverOption) 
  disconnect(): void 
  observe(): void 
  takeRecords(): Report[] 
}
```

File: packages/core/src/domain/report/reportObservable.ts (hunk 6 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:17]: I can also use an alias for the native types that I import. Wdyt?

Comment (BenoitZugmeyer) [reply packages/core/src/domain/report/reportObservable.ts:18]: To avoid conflict, we usually prefix those names with `Rum`: `RumReportType`. WDYT?

File: packages/core/src/domain/report/reportObservable.ts (hunk 5 above)
Comment (bcaudan) [reply packages/core/src/domain/report/reportObservable.ts:15]: what I had in mind: https://github.com/DataDog/browser-sdk/commit/4c522570562a01b4271facaf84873338da6084d0

File: packages/core/src/domain/report/reportObservable.ts (hunk 6 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:18]: I thought that we use `RUM` and `Logs` prefixes when the code is specific to one package like:
`Configuration`, `RumConfiguration`, `LogsConfiguration`

I think `RumReportType` will be misleading. Wdyt?

Comment (bcaudan) [reply packages/core/src/domain/report/reportObservable.ts:18]: Agreed, we use `Raw` prefix at several places, could it fit here?

Comment (BenoitZugmeyer) [reply packages/core/src/domain/report/reportObservable.ts:18]: Ah right, it is used in Logs too. `Raw` seems fine then. In my mind, `Custom` indicates that it is somehow customer related.

File: packages/core/src/domain/report/reportObservable.spec.ts (hunk 8 above)
Comment (bcaudan) [reply packages/core/src/domain/report/reportObservable.spec.ts]: nevermind then

File: packages/core/src/domain/report/reportObservable.ts (hunk 5 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:15]: Way nicer indeed!

File: packages/logs/src/boot/startLogs.spec.ts (hunk 14 above)
Comment (amortemousque) [reply packages/logs/src/boot/startLogs.spec.ts]: Okay, done

File: packages/core/src/domain/report/reportObservable.ts (hunk 6 above)
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:7]: Okay, I'm good with Raw ;)

File: packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.spec.ts [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+  })
+
+  afterEach(() => {
+    subscription.unsubscribe()
+    clock.cleanup()
+    reportingObserverStub.reset()
+    resetExperimentalFeatures()
+  })
+
+  it('should error report when ff forward-reports enabled', () => {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.spec.ts:37]:
```suggestion
  it('should report when ff forward-reports enabled', () => {
```

File: packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+  const subscription = initReportObservable([RawReportType.cspViolation, RawReportType.intervention]).subscribe(
+    (reportError) =>
+      errorObservable.notify({
+        startClocks: clocksNow(),
+        message: reportError.message,
+        stack: reportError.stack,
+        type: reportError.id,
+        source: ErrorSource.REPORT,
+        handling: ErrorHandling.HANDLED,
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts]: Indeed logical

File: packages/core/src/domain/report/reportObservable.ts [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+    stack: buildStack(body.id, body.message, body.sourceFile, body.lineNumber, body.columnNumber),
+  }
+}
+
+function buildRawReportFromCspViolation(event: SecurityPolicyViolationEvent): RawReport {
+  const type = RawReportType.cspViolation
+  const message = `'${event.blockedURI}' blocked by '${event.effectiveDirective}' directive`
+  return {
+    type,
+    id: type,
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts]: Yes it could fit as a subtype.

File: packages/core/src/domain/report/reportObservable.ts [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    type,
+    subtype: body.id,
+    message: `${type}: ${body.message}`,
+    stack: buildStack(body.id, body.message, body.sourceFile, body.lineNumber, body.columnNumber),
+  }
+}
+
+function buildRawReportFromCspViolation(event: SecurityPolicyViolationEvent): RawReport {
+  const type = RawReportType.cspViolation
+  const message = `'${event.blockedURI}' blocked by '${event.effectiveDirective}' directive`
```
Comment (amortemousque) [inline packages/core/src/domain/report/reportObservable.ts:87]: After some thought, I think it is valuable to add the `originalPolicy` to the crafted message. Could be done like so:
`${event.blockedURI}' blocked by '${event.effectiveDirective}' directive of the policy "${event.originalPolicy}"`.

Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:87]: @bcaudan @BenoitZugmeyer

Comment (bcaudan) [reply packages/core/src/domain/report/reportObservable.ts:87]: Some directives could list a large number of domains 🤔
I'd say that it could be added in the stack to keep a minimal message.
We should maybe also truncate the policy to avoid sending too much data.

Comment (amortemousque) [reply packages/core/src/domain/report/reportObservable.ts:87]: I added it to the stack. Let me know what you think

Comment (BenoitZugmeyer) [reply packages/core/src/domain/report/reportObservable.ts:87]: LGTM!

File: packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts (hunk 2 above)
Comment (amortemousque) [reply packages/rum-core/src/domain/rumEventsCollection/error/trackReportError.ts:26]: Done

**Review by bcaudan — APPROVED** [review]
